tokio-util/src/task/spawn_pinned.rs

```diff
@@ -0,0 +1,294 @@
+    /// the worker so the task count is kept consistent with load.
+    fn find_and_incr_least_burdened_worker(&self) -> &LocalWorkerHandle {
+        loop {
+            let (worker, task_count) = self
+                .workers
+                .iter()
+                .map(|worker| (worker, worker.task_count.load(Ordering::SeqCst)))
+                .min_by_key(|&(_, count)| count)
+                .expect("There must be more than one worker");
+
+            // Make sure the task count hasn't changed since when we choose this
+            // worker. Otherwise, restart the search.
+            if worker
+                .task_count
+                .compare_exchange(
+                    task_count,
+                    task_count + 1,
+                    Ordering::SeqCst,
+                    Ordering::Relaxed,
+                )
+                .is_ok()
+            {
+                return worker;
+            }
+        }
+    }
+}
+
+/// Automatically decrements a worker's job count when a job finishes (when
+/// this gets dropped).
+struct JobGuard(Arc<AtomicUsize>);
+
+impl Drop for JobGuard {
+    fn drop(&mut self) {
+        // Decrement the job count
+        self.0.fetch_sub(1, Ordering::SeqCst);
+    }
+}
+
+/// Automatically abort/cancel the task when this guard gets dropped. This will
+/// forward a cancellation from one task to another.
+///
+/// This implements Future by polling the join handle, so just await it.
+struct JoinHandleCancelGuard<T>(JoinHandle<T>);
+
+impl<T> Future for JoinHandleCancelGuard<T> {
+    type Output = <JoinHandle<T> as Future>::Output;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let join_handle = Pin::new(&mut self.0);
+        join_handle.poll(cx)
+    }
+}
+
+impl<T> Drop for JoinHandleCancelGuard<T> {
+    fn drop(&mut self) {
+        // Attempt to abort the task. This does nothing if the task has already
+        // completed.
+        self.0.abort();
+    }
+}
+
+/// If the task is canceled while waiting for the join handle, this guard will
+/// check if the join handle was sent (in-transit so it wasn't aborted on the
+/// worker side) and abort it if so.
+struct ReceiverCancelGuard<T>(oneshot::Receiver<JoinHandle<T>>);
+
+impl<T> Future for ReceiverCancelGuard<T> {
+    type Output = <oneshot::Receiver<JoinHandle<T>> as Future>::Output;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let receiver = Pin::new(&mut self.0);
+        receiver.poll(cx)
+    }
+}
+
+impl<T> Drop for ReceiverCancelGuard<T> {
+    fn drop(&mut self) {
+        // If task is canceled while waiting for the join handle, and the join
+        // handle was already "sent" by the worker, then it's in a limbo state
+        // and needs to be manually canceled here.
+        if let Ok(join_handle) = self.0.try_recv() {
+            join_handle.abort();
+        }
+    }
+}
```

Comment:
This is fragile. We should be able to do better with [`Abortable`][1].

[1]: https://docs.rs/futures/0.3/futures/future/struct.Abortable.html